File: leetcode-go/238.go
package main

func productExceptSelf(nums []int) []int {
	answer := make([]int, len(nums))

	answer[0] = 1
	for i := 1; i < len(nums); i++ {
		answer[i] = answer[i-1] * nums[i-1]
	}

	prev := nums[len(nums)-1]
	for i := len(nums) - 2; i >= 0; i-- {
		answer[i] = answer[i] * prev
		prev *= nums[i]
	}

	return answer
}

// func main() {
// 	arr := []int{1, 2, 3, 4}

// 	productExceptSelf(arr)
// }

File: leetcode-js/238.js
function productExceptSelf(nums) {
    let res = [1];

    for (let i = 1; i < nums.length; i++) {
        res.push(res[i - 1] * nums[i - 1]);
    }

    let acc = 1;
    for (let i = nums.length - 2; i >= 0; i--) {
        res[i] *= acc * nums[i + 1];
        acc *= nums[i + 1];
    }

    return res;
}

let nums = [1, 2, 3, 4];
console.log(productExceptSelf(nums));

